man/man1/ncs.1
'\" t
.\"     Title: ncs
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 01/26/2022
.\"    Manual: NCS Manual
.\"    Source: Cisco Systems, Inc.
.\"  Language: English
.\"
.TH "NCS" "1" "01/26/2022" "Cisco Systems, Inc." "NCS Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ncs \- command to start and control the NCS daemon
.SH "SYNOPSIS"
.HP \w'\fBncs\fR\ 'u
\fBncs\fR [\-\-conf\ \fIConfFile\fR] [\-\-cd\ \fIDir\fR] [\-\-addloadpath\ \fIDir\fR] [\-\-nolog] [\-\-smp\ \fINr\fR] [\-\-foreground\ [\-v\ |\ \-\-verbose]\ [\-\-stop\-on\-eof]] [\-\-with\-package\-reload] [\-\-ignore\-initial\-validation] [\-\-full\-upgrade\-validation] [\-\-start\-phase0] [\-\-epoll\ {true\ |\ false}]
.HP \w'\fBncs\fR\ 'u
\fBncs\fR {\-\-wait\-phase0\ [\ \fITryTime\fR]  | \-\-start\-phase1 | \-\-start\-phase2 | \-\-wait\-started\ [\ \fITryTime\fR]  | \-\-reload | \-\-areload | \-\-status | \-\-check\-callbacks\ [\fINamespace\fR\ |\ \fIPath\fR] | \-\-loadfile\ \fIFile\fR | \-\-rollback\ \fINr\fR | \-\-debug\-dump\ \fIFile\fR | \-\-cli\-j\-dump\ \fIFile\fR | \-\-loadxmlfiles\ \fIFile\fR... | \-\-mergexmlfiles\ \fIFile\fR... | \-\-stop} [\-\-timeout\ \fIMaxTime\fR]
.HP \w'\fBncs\fR\ 'u
\fBncs\fR {\-\-version | \-\-cdb\-debug\-dump\ \fIDirectory\fR}
.SH "DESCRIPTION"
.PP
Use this command to start and control the NCS daemon\&.
.SH "STARTING NCS"
.PP
These options are relevant when starting the NCS daemon\&.
.PP
\fB\-c\fR, \fB\-\-conf\fR ConfFile
.RS 4
ConfFile is the path to a ncs\&.conf file\&. If the
\-c File
argument is not given to
ncs, first
$PWD
is searched for a file called
ncs\&.conf, if not found
$NCS_DIR/etc/ncs/ncs\&.conf
is chosen\&.
.RE
.PP
\fB\-\-cd\fR Dir
.RS 4
Change working directory
.RE
.PP
\fB\-\-addloadpath\fR Dir
.RS 4
Add Dir to the set of directories NCS uses to load fxs, clispec, NCS packages and SNMP bin files\&.
.RE
.PP
\fB\-\-nolog\fR
.RS 4
Do not log initial startup messages to syslog\&.
.RE
.PP
\fB\-\-smp\fR Nr
.RS 4
Number of threads to run for Symmetric Multiprocessing (SMP)\&. The default is to enable SMP support, with as many threads as the system has logical processors, if more than one logical processor is detected\&. Giving a value of 1 will disable SMP support, while a value bigger than 1 will enable SMP support, where NCS will at any given time use at most as many logical processors as the number of threads\&.
.RE
.PP
\fB\-\-foreground [ \-v | \-\-verbose ] [ \-\-stop\-on\-eof ] \fR
.RS 4
Do not start as a daemon\&. Can be used to start NCS from a process manager\&. In combination with \-v or \-\-verbose, all log messages are printed to stdout\&. Useful during development\&. In combination with \-\-stop\-on\-eof, NCS will stop if it receives EOF (ctrl\-d) on standard input\&. Note that to stop NCS when run in foreground, send EOF (if \-\-stop\-on\-eof was used) or use ncs \-\-stop\&. Do not terminate with ctrl\-c, since NCS in that case won\*(Aqt have the chance to close the database files\&.
.RE
.PP
\fB\-\-with\-package\-reload\fR
.RS 4
When NCS starts, if the private package directory tree already exists, NCS will load the packages from this directory tree and not search the load\-path for packages\&. If the \-\-with\-package\-reload option is given when starting NCS, the load\-path will be searched and the packages found there copied to the private package directory tree, replacing the previous contents, before loading\&. This should always be used when upgrading to a new version of NCS in an existing directory structure, to make sure that new packages are loaded together with the other parts of the new system\&.
.sp
When NCS is started from the /etc/init\&.d scripts, that get generated by the \-\-system\-install option to the NCS installer, the environment variable NCS_RELOAD_PACKAGES can be set to \*(Aqtrue\*(Aq to attempt a package reload\&.
.RE
.PP
\fB\-\-with\-package\-reload\-force\fR
.RS 4
When reloading packages NCS will give a warning when the upgrade looks "suspicious", i\&.e\&. may break some functionality\&. This is not a strict upgrade validation, but only intended as a hint to NSO administrator early in the upgrade process that something might be wrong\&. Please refer to Loading Packages section in NSO Administration Guide for more information\&.
.sp
If all changes indicated by the warnings are intended, this option allows to override warnings and proceed with the upgrade\&. This option is equivalent to setting NCS_RELOAD_PACKAGES environment variable to \*(Aqforce\*(Aq\&.
.RE
.PP
\fB\-\-ignore\-initial\-validation\fR
.RS 4
When CDB starts on an empty database, or when upgrading, it starts a transaction to load the initial configuration or perform the upgrade\&. This option makes NCS skip any validation callpoints when committing these initial transaction\&. (The preferred alternative is to use start\-phases and register the validation callpoints in phase 0, see the user guide)\&.
.RE
.PP
\fB\-\-full\-upgrade\-validation\fR
.RS 4
Perform a full validation of the entire database if the data models have been upgraded\&. This is useful in order to trigger external validation to run even if the database content has not been modified\&.
.RE
.PP
\fB\-\-start\-phase0\fR
.RS 4
Start the daemon, but only start internal subsystems and CDB\&. Phase 0 is used when a controlled upgrade is done\&.
.RE
.PP
\fB\-\-epoll { true | false }\fR
.RS 4
Determines whether NCS should use an enhanced poll() function (e\&.g\&. Linux epoll(7))\&. This can improve performance when NCS has a high number of connections, but there may be issues with the implementation in some OS/kernel versions\&. The default is true\&.
.RE
.SH "COMMUNICATING WITH NCS"
.PP
When the NCS daemon has been started, these options are used to communicate with the running daemon\&.
.PP
By default these options will perform their function by connecting to a running NCS daemon over the loopback interface on the standard port\&. If the environment variable
\fBNCS_IPC_PORT\fR
and/or
\fBNCS_IPC_ADDR\fR
are set then the address/port in those variables will be used to communicate with the NCS daemon\&. (Must be used if the daemon is not listening on its standard port on localhost, see the
/ncs\-config/ncs\-ipc\-address/ip
setting in the
\m[blue]\fBncs\&.conf(5)\fR\m[]
man\-page, and the section on
\m[blue]\fBNCS IPC\fR\m[]
in the NCS Users Guide)\&.
.PP
\fB\-\-wait\-phase0 [ TryTime ]\fR
.RS 4
This call hangs until NCS has initialized start phase0\&. After this call has returned, it is safe to register validation callbacks, upgrade CDB etc\&. This function is useful when NCS has been started with \-\-foreground and \-\-start\-phase0\&. It will keep trying the initial connection to NCS for at most TryTime seconds (default 5)\&.
.RE
.PP
\fB\-\-start\-phase1\fR
.RS 4
Do not start the subsystems that listen to the management IP address\&. Must be called after the daemon was started with \-\-start\-phase0\&.
.RE
.PP
\fB\-\-start\-phase2\fR
.RS 4
Must be called after the management interface has been brought up, if \-\-start\-phase1 has been used\&. Starts the subsystems that listens to the management IP address\&.
.RE
.PP
\fB\-\-wait\-started [ TryTime ]\fR
.RS 4
This call hangs until NCS is completely started\&. This function is useful when NCS has been started with \-\-foreground\&. It will keep trying the initial connection to NCS for at most TryTime seconds (default 5)\&.
.RE
.PP
\fB\-\-reload\fR
.RS 4
Reload the NCS daemon configuration\&. All log files are closed and reopened, which means that
\fBncs \-\-reload\fR
can be used from e\&.g\&. logrotate(8), but it is more efficient to use
\fBncs_cmd \-c reopen_logs\fR
for this purpose\&. Note: If we update a \&.fxs file it is not enough to do a reload; the "packages reload" action must be invoked, or the daemon must be restarted with the
\fB\-\-with\-package\-reload\fR
option\&.
.RE
.PP
\fB\-\-areload\fR
.RS 4
Asynchronously reload the NCS daemon configuration\&. This can be used in scripts executed by the NCS daemon\&.
.RE
.PP
\fB\-\-stop\fR
.RS 4
Stop the NCS daemon\&.
.RE
.PP
\fB\-\-status\fR
.RS 4
Prints status information about the NCS daemon on stdout\&. Among the things listed are: loaded namespaces, current user sessions, callpoints (and whether they are registered or not), CDB status, and the current start\-phase\&. Start phases are reported as "status:" and can be one of starting (which is pre\-phase0), phase0, phase1, started (i\&.e\&. phase2), or stopping (which means that NCS is about to shutdown)\&.
.RE
.PP
\fB\-\-debug\-dump File\fR
.RS 4
Dump debug information from an already running NCS daemon into a file\&. The file only makes sense to NCS developers\&. It is often a good idea to include a debug dump in NCS trouble reports\&.
.RE
.PP
\fB\-\-cli\-j\-dump File\fR
.RS 4
Dump cli structure information from the NCS daemon into a file\&.
.RE
.PP
\fB\-\-check\-callbacks [Namespace | Path]\fR
.RS 4
Walks through the entire data tree (config and stat), or only the Namespace or Path, and verifies that all read\-callbacks are implemented for all elements, and verifies their return values\&.
.RE
.PP
\fB\-\-loadfile File\fR
.RS 4
Load configuration in curly bracket format from File\&.
.RE
.PP
\fB\-\-rollback Nr\fR
.RS 4
Rollback configuration to saved configuration number Nr\&.
.RE
.PP
\fB\-\-loadxmlfiles File \&.\&.\&.\fR
.RS 4
Load configuration in XML format from Files\&. The configuration is completely replaced by the contents in Files\&.
.RE
.PP
\fB\-\-mergexmlfiles File \&.\&.\&.\fR
.RS 4
Load configuration in XML format from Files\&. The configuration is merged with the contents in Files\&. The XML may use the \*(Aqoperation\*(Aq attribute, in the same way as it is used in a NETCONF <edit\-config> operation\&.
.RE
.PP
\fB\-\-timeout MaxTime\fR
.RS 4
Specify the maximum time to wait for the NCS daemon to complete the command, in seconds\&. If this option is not given, no timeout is used\&.\&.
.RE
.SH "STANDALONE OPTIONS"
.PP
\fB\-\-cdb\-debug\-dump Directory\fR
.RS 4
Print a lot of information about the CDB files in
\fIDirectory\fR
to stdout\&. This is a completely stand\-alone feature and the only thing needed is the \&.cdb files (no running NCS daemon or \&.fxs files etc)\&.
.sp
It is possible to check CDB files for invalid UTF\-8 content by appending the
\fIvalidate_utf8\fR
option after the
\fIDirectory\fR
like so:
.sp
\fBncs \-\-cdb\-debug\-dump Directory validate_utf8\fR
.sp
This will emit all paths and content with invalid UTF\-8\&.
.RE
.PP
\fB\-\-version\fR
.RS 4
Reports the ncs version without interacting with the daemon\&.
.RE
.PP
\fB\-\-timeout MaxTime\fR
.RS 4
See above
.RE
.SH "ENVIRONMENT"
.PP
When NCS is started from the /etc/init\&.d scripts, that get generated by the \-\-system\-install option to the NCS installer, the environment variable NCS_RELOAD_PACKAGES can be set to \*(Aqtrue\*(Aq to attempt a package reload\&.
.SH "DIAGNOSTICS"
.PP
If NCS starts, the exit status is 0\&. If not it is a positive integer\&. The different meanings of the different exit codes are documented in the "NCS System Management" chapter in the user guide\&. When failing to start, the reason is stated in the NCS daemon log\&. The location of the daemon log is specified in the ConfFile as described in
\m[blue]\fBncs\&.conf(5)\fR\m[]\&.
.SH "SEE ALSO"
.PP
\fBncs.conf\fR(5)
\- NCS daemon configuration file format
.SH "AUTHOR"
.br
.RS 4
Author.
.RE
.SH "COPYRIGHT"
.br
Copyright \(co 2021, 2022 Cisco Systems, Inc. All rights reserved.
.br


CHANGES
NSO-5.7.1 [January 26, 2022]
============================
Security:

  - java-api: Security fix. This vulnerability affects NSO releases 5.3
    and later.

    (ENG-28001)

Enhancements:

  - ncs: Device locks (implemented as commit queue locks) now have a more
    descriptive tag. This tag is visible in status data and notifications.

    (ENG-27962)

  - ncs: Service actions (deep-check-sync and get-modifications) work
    correctly in 3-layer LSA.

    (ENG-24354)

  - cs_trans: Skip evaluating "when" expression for a leaf without a default
    value when a parent node is created, since the leaf does not have a
    value yet.

    (ENG-28019)

  - java-api: log4j has been upgraded to version 2.17.1

    (ENG-28036, CSCwa54664, CSCwa54674)

  - observability: The "ncs --debug-dump <File>" command now supports
    additional options:
    (1) --collect-timeout <seconds> to provide the desired timeout when
    system collects internal data. The default timeout is 10 seconds.
    (2) --compress to compress the debug dump to File.gz.
    Additionally, if a timeout is reached, NSO will inform the client with
    the following message:
    *** Failed to collect <type>: timeout

    (ENG-27926, RT:47738, PS-42861, SR:692398999)

Corrections:

  - ncs: The NED migration no longer sets service meta-data attributes on
    leaf-lists that have been deleted on the device, causing a broken CDB
    journal.

    (ENG-28012, RT:47926, PS-43038, CSCwa52384, SR:692520230)

  - ncs: Fixed incorrect change-set handling that would sometimes lead to
    schema violations in the write-ahead log. This caused data nodes to
    appear in the tree with the wrong schema or new nodes being created
    during delete operations.

    (ENG-26915, RT:47668, RT:45996, RT:47442, PS-42790, PS-41113, PS-42563)

  - ncs: Filtering progress trace events based only on context or user did
    not work correctly.

    (ENG-26901, RT:46498, PS-41612, CSCvy73354, SR:691550999)

  - candidate: Fixed wrongly removing inactive attribute when committing
    candidate configuration.

    (ENG-27710)

  - cdb: Corrected set_case() callback in operational datastore when
    applying nested 'choice' statements.

    (ENG-24931)

  - cdb-capi: A close event from a CDB subscriber socket while processing
    cdb_oper write requests is now handled properly instead of restarting
    NSO.

    (ENG-27531, RT:47665, PS-42787, CSCwa00963)

  - cdb-capi: cdb_get_txid() wrongly returned 'Unknown CDB operation' on a
    CDB session which was established at the prepare stage for a read-write
    transaction. The CDB has been updated and will now correctly return the
    transaction ID.

    (ENG-25840, RT:45363, PS-40477, CSCvx61089)

  - cli: When traversing the command line to execute, the candidate node
    matching algorithm used prefixes even when it was not required. This
    resulted in syntax errors when the command did not include a prefix, and
    has been corrected.

    (ENG-27723, RT:47882, PS-42994, CSCwa21504)

  - cs_trans: Fixed "when" expression not evaluating if it depends on a leaf
    with a default value, the leaf is removed, then set to the old value
    again in the same transaction.

    (ENG-27931)

  - cs_trans: Corrected a fatal "bad dependency" error  occurring with a
    "must" expression on a leaf-list that refers to a list node ancestor.

    (ENG-27917, RT:48094, PS-43209, CSCwa40711)

  - cs_trans: In a nested choice node, the subscription notification for the
    default leaf under nested default case was missing when deleting the
    non-default leaf in the same choice. Now, the subscription notification
    will be present for the default leaf under nested default case when
    deleting the non-default leaf in the same choice.

    (ENG-27123, RT:47119, PS-42238, CSCvz42203)

  - ha: Removed a race condition when pending rollback files where aborted
    and deleted just before HA tried to distribute the file to the
    secondaries. The race condition resulted in an enoent - file not found
    error.

    (ENG-27882)

  - netconf: Filter from NETCONF notification modify-subscription operation
    was not propagated into the monitor structure /sn:subscriptions. Now the
    filter is properly stored for a modify-subscription request and can be
    read from /sn:subscriptions afterwards.

    (ENG-27610, RT:47752, PS-42875, CSCwa09847, SR:692301317)

  - netconf: An ongoing NETCONF notification establish-subscription, that
    was abruptly terminated by killing the user session, sometimes resulted
    in a zombie subscription being left in the monitor model
    /sn:subscriptions in the CDB. A following attempt of an
    establish-subscription would fail due to clash with the leftover zombie
    subscription.
    To avoid zombie subscriptions in /sn:subscriptions, an internal data
    provider (DP) now handles client reads from /sn:subscriptions. The DP
    also protects /sn:subscriptions from being written to by clients, as
    only dynamic subscriptions are supported.

    (ENG-27115, RT:46541, PS-41655, CSCvz39935)

  - netconf: Add status for the running database in netconf-state, as well
    as the corresponding set_running_status and get_running_status commands
    for the ncs_cmd utility.
    Change of the status is logged in the developer log.

    (ENG-26042, RT:45616, PS-40734, CSCvx84625)

  - restconf: When using Yang-Patch with JSON payloads, the order in which
    list keys are presented is now truly irrelevant.

    (ENG-27423, RT:47554, PS-42675, CSCvz86603, SR:692230863)

  - upgrade: The upgrade process incorrectly handled an operational node
    with 'tailf:persistent true' statement being changed to
    'tailf:persistent false' and, therefore, a system restart did not clear
    the node's value in the datastore. The upgrade process has been fixed
    and the node will be removed after system restart.

    (ENG-27180, RT:45088, PS-40201, CSCvz52132)


src/ncs/package-skeletons/etc/suse/ncs.init.d
#! /bin/sh
#
# SuSE Startup script for NCS
#
### BEGIN INIT INFO
# Provides:          ncs
# Required-Start:    $local_fs $network $syslog
# Required-Stop:
# X-Start-Before:
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: NCS
# Description: NCS - Network Control System
### END INIT INFO

ncsdir=%ncsdir%
confdir=%confdir%
rundir=%rundir%
logdir=%logdir%

ncs=${ncsdir}/bin/ncs
prog=ncs
conf="-c ${confdir}/ncs.conf"
heart="--heart"


test -x $ncs || exit 5


# Shell functions sourced from /etc/rc.status:
#      rc_check         check and set local and overall rc status
#      rc_status        check and set local and overall rc status
#      rc_status -v     ditto but be verbose in local rc status
#      rc_status -v -r  ditto and clear the local rc status
#      rc_failed        set local and overall rc status to failed
#      rc_failed <num>  set local and overall rc status to <num><num>
#      rc_reset         clear local rc status (overall remains)
#      rc_exit          exit appropriate to overall rc status
#      rc_active        checks whether a service is activated by symlinks
. /etc/rc.status

# First reset status of this service
rc_reset

# Return values acc. to LSB for all commands but status:
# 0 - success
# 1 - generic or unspecified error
# 2 - invalid or excess argument(s)
# 3 - unimplemented feature (e.g. "reload")
# 4 - insufficient privilege
# 5 - program is not installed
# 6 - program is not configured
# 7 - program is not running
#
# Note that starting an already running service, stopping
# or restarting a not-running service as well as the restart
# with force-reload (in case signalling is not supported) are
# considered a success.

case "$1" in
    start)
        echo -n "Starting ${prog}"
        . $ncsdir/ncsrc
        NCS_CONFIG_DIR=${confdir}
        NCS_RUN_DIR=${rundir}
        NCS_LOG_DIR=${logdir}
        export NCS_CONFIG_DIR NCS_RUN_DIR NCS_LOG_DIR
        %sustart%${ncs} --cd ${rundir} ${heart} ${conf}%suend%

        # Remember status and be verbose
        rc_status -v
        ;;
    stop)
        echo -n "Shutting down ${prog}"
        ${ncs} --stop

        # Remember status and be verbose
        rc_status -v
        ;;
    try-restart)
        ## Stop the service and if this succeeds (i.e. the
        ## service was running before), start it again.
        ## Note: try-restart is not (yet) part of LSB (as of 0.7.5)
        $0 status >/dev/null &&  $0 restart

        # Remember status and be quiet
        rc_status
        ;;
    restart)
        ## Stop the service and regardless of whether it was
        ## running or not, start it again.
        $0 stop
        $0 start

        # Remember status and be quiet
        rc_status
        ;;
    force-reload)
        echo -n "Force Reload service ${prog}"
        ## if it supports it:
        $ncs --reload
        rc_status -v
        ;;
    reload)
        ## Like force-reload, but if daemon does not support
        ## signalling, do nothing (!)
        # If it supports signalling:
        $ncs --reload
        rc_status -v
        ;;
    status)
        echo -n "Checking for service ${prog}: "
        $ncs --status
        rc_status -v
        ;;
    probe)
        ## Optional: Probe for the necessity of a reload,
        ## print out the argument which is required for a reload.
        test /etc/FOO/FOO.conf -nt /var/run/FOO.pid && echo reload
        ;;
    start-with-package-reload)
        ## Start NCS and attempt an NCS package code reload
        NCS_RELOAD_PACKAGES=true $0 start
        rc_status
        ;;
    restart-with-package-reload)
        ## Restart NCS and attempt an NCS package code reload
        NCS_RELOAD_PACKAGES=true $0 restart
        rc_status
        ;;
    *)
        echo "Usage: $0 {start|stop|status|try-restart|restart|force-reload|reload|probe}"
        exit 1
        ;;
esac
rc_exit
